src/firefly/js/core/ReduxFlux.js

```diff
@@ -96,238 +12,72 @@ import ImageLineBasedFootprint from '../drawingLayers/ImageLineBasedFootprint.js
- * payload parameters.  The utility function should implement the following standard:
- * <ul>
- *     <li>The function name should start with "dispatch"</li>
- *     <li>The action type as the second part of the name</li>
- *     <li>The function should be exported from the controller</li>
- *     <li>The function parameters should the documented with jsdocs</li>
- *     <li>Optional parameters should be clear</li>
- * </ul>
- * Utility function Example - if action type is <code>PLOT_IMAGE</code> and the <code>PLOT_IMAGE</code> action
- * is exported from the ImagePlotCntlr module.  The the name should be <code>processPlotImage</code>.
- *
- *
- * @param {Action} rawAction
- */
-function process(rawAction) {
-    if (!redux) throw Error('firefly has not been bootstrapped');
-
-    getOrCreateWsConn().catch(() => showLostConnection());
-    var ac = actionCreators.get(rawAction.type);
-    if (!rawAction.payload) rawAction= Object.assign({},rawAction,{payload:{}});
-    if (ac) {
-        redux.dispatch(ac(rawAction));
-    } else {
-        redux.dispatch( rawAction );
+function makeProcess(redux, getActionCreators, preDispatch, postDispatch) {
+    /**
+     * Process the rawAction.  This uses the actionCreators map to resolve
+     * the ActionCreator given the action.type.  If one is not mapped, then it'll
+     * create a simple 'pass through' ActionCreator that returns the rawAction as an action.
+     *
+     * <i>Note: </i> Often it makes sense to have a utility function call <code>process</code>. In that case
+     * the utility function should meet the follow criteria.  This is a good way to document and default the
+     * payload parameters.  The utility function should implement the following standard:
+     * <ul>
+     *     <li>The function name should start with "dispatch"</li>
+     *     <li>The action type as the second part of the name</li>
+     *     <li>The function should be exported from the controller</li>
+     *     <li>The function parameters should the documented with jsdocs</li>
+     *     <li>Optional parameters should be clear</li>
+     * </ul>
+     * Utility function Example - if action type is <code>PLOT_IMAGE</code> and the <code>PLOT_IMAGE</code> action
+     * is exported from the ImagePlotCntlr module.  The the name should be <code>processPlotImage</code>.
+     *
+     *
+     * @param {Action} rawAction
+     */
+    const process= (rawAction) => {
+        preDispatch();
+        const ac = getActionCreators().get(rawAction.type);
+        if (!rawAction.payload) rawAction= {...rawAction, payload:{}};
+        redux.dispatch(ac ? ac(rawAction) : rawAction);
+        postDispatch(rawAction);
+    };
+    return process;
+}
+
+export function bootstrapRedux(bootstrapRegistry, preDispatch= () => undefined, postDispatch= () => undefined) {
+    if (!bootstrapRegistry) {
+        Logger('ReduxFlux').error('bootstrapRegistry parameter is require for firefly startup');
+        return;
```

Comment:
I think allowing Firefly to change the behavior of `process` is a great idea.  I don't really like how `pre/post` is passed from here, all the way to `makeProcess`.
How about, instead of `pre/post` we take a decorator function.  We will apply it if exists.  i.e.

```
export function bootstrapRedux(bootstrapRegistry, processDecor= (process) => process ) {
    const process = processDecor(makeProcess(redux,bootstrapRegistry.getActionCreators)) : 
}  
```

This way, `makeProcess` does only what it suppose without external logic.

Reply:
sure, the decorator idea did not occur to me. I will do that.